File: python/ldap_filter.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import csv
import io
import json
import lzma
import re
import sys
import time
from dataclasses import dataclass, field
from typing import Iterable, Optional, TextIO


_MISSING = object()
_ITEM_RE = re.compile(r"\A([^~=><]+)(~=|>=|<=|=)(.+)\Z")
_SYMBOL_NAME_RE = re.compile(r"\A[A-Za-z_][A-Za-z0-9_]*\Z")


class LdapFilterError(Exception):
    pass


@dataclass
class OrderedAttrs:
    items: list[tuple[str, Optional[str]]] = field(default_factory=list)

    def add(self, key: str, value: Optional[str]) -> None:
        self.items.append((key, value))

    def contains(self, key: str) -> bool:
        return any(candidate == key for candidate, _ in self.items)

    def get(self, key: str):
        for candidate, value in self.items:
            if candidate == key:
                return value
        return _MISSING


@dataclass(frozen=True)
class ItemMatcher:
    kind: str
    parts: tuple[str, ...] = ()
    leading_star: bool = False
    trailing_star: bool = False


class FilterExpr:
    def evaluate(self, attrs: OrderedAttrs) -> bool:
        raise NotImplementedError


@dataclass(frozen=True)
class FilterAnd(FilterExpr):
    children: tuple[FilterExpr, ...]

    def evaluate(self, attrs: OrderedAttrs) -> bool:
        return all(child.evaluate(attrs) for child in self.children)


@dataclass(frozen=True)
class FilterOr(FilterExpr):
    children: tuple[FilterExpr, ...]

    def evaluate(self, attrs: OrderedAttrs) -> bool:
        return any(child.evaluate(attrs) for child in self.children)


@dataclass(frozen=True)
class FilterNot(FilterExpr):
    child: FilterExpr

    def evaluate(self, attrs: OrderedAttrs) -> bool:
        return not self.child.evaluate(attrs)


@dataclass(frozen=True)
class FilterItem(FilterExpr):
    attr: str
    op: str
    value: str
    matcher: ItemMatcher

    def evaluate(self, attrs: OrderedAttrs) -> bool:
        actual = attrs.get(self.attr)

        if self.op == "=":
            if self.matcher.kind == "presence":
                return attrs.contains(self.attr)
            if actual is _MISSING:
                return False
            if self.matcher.kind == "wildcard":
                return isinstance(actual, str) and wildcard_matches(
                    self.matcher.parts,
                    self.matcher.leading_star,
                    self.matcher.trailing_star,
                    actual,
                )
            return actual == self.value

        if actual is _MISSING or actual is None:
            return False

        if self.op == "~=":
            return levenshtein_distance_lte(str(actual), self.value, 2)
        if self.op == ">=":
            return str(actual) >= self.value
        if self.op == "<=":
            return str(actual) <= self.value
        raise LdapFilterError(f"unsupported operator: {self.op}")


class FilterParser:
    def __init__(self, text: str):
        self.text = text
        self.pos = 0

    def parse(self) -> FilterExpr:
        expr = self._parse_filter()
        if self.pos != len(self.text):
            raise self._error("unexpected trailing characters")
        return expr

    def _parse_filter(self) -> FilterExpr:
        self._expect("(")
        current = self._peek()
        if current is None:
            raise self._error("parenthesis mismatch")

        if current == "&":
            self.pos += 1
            children = self._parse_subfilters()
            if not children:
                raise self._error("and operator requires filters")
            self._expect(")")
            return FilterAnd(tuple(children))

        if current == "|":
            self.pos += 1
            children = self._parse_subfilters()
            if not children:
                raise self._error("or operator requires filters")
            self._expect(")")
            return FilterOr(tuple(children))

        if current == "!":
            self.pos += 1
            child = self._parse_filter()
            if self._peek() == "(":
                raise self._error("not operator has more than one filter")
            self._expect(")")
            return FilterNot(child)

        item = self._parse_item()
        self._expect(")")
        return item

    def _parse_subfilters(self) -> list[FilterExpr]:
        children: list[FilterExpr] = []
        while self._peek() == "(":
            children.append(self._parse_filter())
        return children

    def _parse_item(self) -> FilterItem:
        start = self.pos
        while True:
            current = self._peek()
            if current is None:
                raise self._error("parenthesis mismatch")
            if current == ")":
                break
            self.pos += 1

        item = self.text[start:self.pos]
        match = _ITEM_RE.match(item)
        if match is None:
            raise self._error("error in item syntax", start)

        attr = match.group(1)
        op = match.group(2)
        raw_value = match.group(3)
        value, matcher = parse_item_value(raw_value, start)
        return FilterItem(attr=attr, op=op, value=value, matcher=matcher)

    def _peek(self) -> Optional[str]:
        if self.pos >= len(self.text):
            return None
        return self.text[self.pos]

    def _expect(self, expected: str) -> None:
        actual = self._peek()
        if actual != expected:
            if actual is None:
                raise self._error(f"expected {expected!r}, found EOF")
            raise self._error(f"expected {expected!r}, found {actual!r}")
        self.pos += 1

    def _error(self, message: str, position: Optional[int] = None) -> LdapFilterError:
        if position is None:
            position = self.pos
        return LdapFilterError(f"{message} at position {position}")


def parse_filter(text: str) -> FilterExpr:
    parser = FilterParser(text)
    return parser.parse()


def parse_item_value(raw: str, position: int) -> tuple[str, ItemMatcher]:
    if raw == "*":
        return "*", ItemMatcher(kind="presence")

    decoded: list[str] = []
    segments: list[str] = []
    current: list[str] = []
    saw_wildcard = False
    i = 0

    while i < len(raw):
        ch = raw[i]
        if ch == "\\":
            if i + 2 >= len(raw):
                raise LdapFilterError(f"incomplete escape sequence at position {position}")
            decoded_char = decode_hex_escape(raw[i + 1], raw[i + 2], position)
            decoded.append(decoded_char)
            current.append(decoded_char)
            i += 3
            continue
        if ch == "*":
            saw_wildcard = True
            decoded.append("*")
            segments.append("".join(current))
            current = []
            i += 1
            continue

        decoded.append(ch)
        current.append(ch)
        i += 1

    decoded_text = "".join(decoded)
    if not saw_wildcard:
        return decoded_text, ItemMatcher(kind="exact")

    segments.append("".join(current))
    return decoded_text, ItemMatcher(
        kind="wildcard",
        parts=tuple(segments),
        leading_star=raw.startswith("*"),
        trailing_star=raw.endswith("*"),
    )


def decode_hex_escape(high: str, low: str, position: int) -> str:
    try:
        return chr(int(high + low, 16))
    except ValueError as exc:
        raise LdapFilterError(f"invalid escape sequence: \\{high}{low} at position {position}") from exc


def wildcard_matches(parts: Iterable[str], leading_star: bool, trailing_star: bool, actual: str) -> bool:
    non_empty = [part for part in parts if part]
    if not non_empty:
        return True

    position = 0
    for index, part in enumerate(non_empty):
        is_first = index == 0
        is_last = index == len(non_empty) - 1

        if is_first and not leading_star:
            if not actual[position:].startswith(part):
                return False
            position += len(part)
            continue

        if is_last and not trailing_star:
            return actual[position:].endswith(part)

        found = actual.find(part, position)
        if found < 0:
            return False
        position = found + len(part)

    return True


def levenshtein_distance_lte(a: str, b: str, max_distance: int) -> bool:
    a_chars = list(a)
    b_chars = list(b)
    if abs(len(a_chars) - len(b_chars)) > max_distance:
        return False

    prev = list(range(len(b_chars) + 1))
    curr = [0] * (len(b_chars) + 1)

    for i, a_ch in enumerate(a_chars):
        curr[0] = i + 1
        row_min = curr[0]

        for j, b_ch in enumerate(b_chars):
            cost = 0 if a_ch == b_ch else 1
            delete = prev[j + 1] + 1
            insert = curr[j] + 1
            substitute = prev[j] + cost
            value = min(delete, insert, substitute)
            curr[j + 1] = value
            row_min = min(row_min, value)

        if row_min > max_distance:
            return False

        prev, curr = curr, prev

    return prev[-1] <= max_distance


def escape_ruby_string(text: str) -> str:
    return json.dumps(text)[1:-1]


def format_key(key: str) -> str:
    if _SYMBOL_NAME_RE.match(key):
        return f"{key}: "
    return f"{json.dumps(key)} => "


def format_value(value: Optional[str]) -> str:
    if value is None:
        return "nil"
    return f"\"{escape_ruby_string(str(value))}\""


def inspect_attrs(attrs: OrderedAttrs) -> str:
    parts = [f"{format_key(key)}{format_value(value)}" for key, value in attrs.items]
    return "{" + ", ".join(parts) + "}"


def parse_csv_header(line: str) -> list[str]:
    return parse_csv_line(line.lstrip("\ufeff"))


def parse_csv_line(line: str) -> list[str]:
    try:
        reader = csv.reader([line])
        return next(reader)
    except (csv.Error, StopIteration):
        return [line]


def row_to_attrs(headers: list[str], row: list[str]) -> OrderedAttrs:
    attrs = OrderedAttrs()
    for index, header in enumerate(headers):
        value = row[index] if index < len(row) else ""
        attrs.add(header, value)
    return attrs


def unescape_ltsv_value(value: str) -> Optional[str]:
    if value == "":
        return None

    out: list[str] = []
    i = 0
    while i < len(value):
        ch = value[i]
        if ch != "\\":
            out.append(ch)
            i += 1
            continue

        if i + 1 >= len(value):
            out.append("\\")
            break

        escaped = value[i + 1]
        if escaped == "r":
            out.append("\r")
        elif escaped == "n":
            out.append("\n")
        elif escaped == "t":
            out.append("\t")
        elif escaped == "\\":
            out.append("\\")
        else:
            out.append("\\")
            out.append(escaped)
        i += 2

    result = "".join(out)
    return None if result == "" else result


def parse_ltsv_line(line: str) -> OrderedAttrs:
    attrs = OrderedAttrs()
    if line == "":
        return attrs

    for entry in line.split("\t"):
        if ":" not in entry:
            continue
        key, value = entry.split(":", 1)
        attrs.add(key, unescape_ltsv_value(value))
    return attrs


def detect_format(input_path: str) -> str:
    lower = input_path.lower()
    if lower.endswith(".csv") or lower.endswith(".csv.xz"):
        return "csv"
    if lower.endswith(".ltsv") or lower.endswith(".ltsv.xz"):
        return "ltsv"
    return "ltsv"


def phase_line(phase: str, started_ns: int) -> str:
    elapsed_ns = time.perf_counter_ns() - started_ns
    return f"phase={phase} t={elapsed_ns} elapsed_ns={elapsed_ns}"


def open_input(path: str):
    if path.lower().endswith(".xz"):
        return lzma.open(path, "rt", encoding="utf-8", newline="")
    return open(path, "rt", encoding="utf-8", newline="")


def run_with_io(argv: list[str], stdout: TextIO = sys.stdout, stderr: TextIO = sys.stderr) -> int:
    started_ns = time.perf_counter_ns()
    stderr.write(phase_line("boot", started_ns) + "\n")

    parser = argparse.ArgumentParser(prog="ldap_filter.py")
    parser.add_argument("--format", choices=["auto", "csv", "ltsv"], default="auto")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--jit", action="store_true")
    group.add_argument("--no-jit", action="store_true")
    parser.add_argument("filter")
    parser.add_argument("input")

    try:
        args = parser.parse_args(argv)
        format_kind = args.format if args.format != "auto" else detect_format(args.input)
        expr = parse_filter(args.filter)

        with open_input(args.input) as handle:
            if format_kind == "csv":
                header_line = handle.readline()
                headers = parse_csv_header(header_line.rstrip("\r\n")) if header_line else []
            else:
                headers = []

            stderr.write(phase_line("ready", started_ns) + "\n")

            if format_kind == "csv":
                for line in handle:
                    row = parse_csv_line(line.rstrip("\r\n"))
                    attrs = row_to_attrs(headers, row)
                    if expr.evaluate(attrs):
                        stdout.write(inspect_attrs(attrs) + "\n")
            else:
                for line in handle:
                    attrs = parse_ltsv_line(line.rstrip("\r\n"))
                    if expr.evaluate(attrs):
                        stdout.write(inspect_attrs(attrs) + "\n")

        stderr.write(phase_line("done", started_ns) + "\n")
        return 0
    except LdapFilterError as exc:
        stderr.write(str(exc) + "\n")
        return 1
    except OSError as exc:
        stderr.write(str(exc) + "\n")
        return 1


def main(argv: Optional[list[str]] = None) -> int:
    if argv is None:
        argv = sys.argv[1:]
    return run_with_io(argv)


if __name__ == "__main__":
    raise SystemExit(main())

File: php/ldap_filter.php
<?php
declare(strict_types=1);

final class LdapFilterError extends RuntimeException
{
}

abstract class FilterExpr
{
    abstract public function evaluate(OrderedAttrs $attrs): bool;
}

final class FilterAnd extends FilterExpr
{
    /**
     * @param FilterExpr[] $children
     */
    public function __construct(private array $children)
    {
    }

    public function evaluate(OrderedAttrs $attrs): bool
    {
        foreach ($this->children as $child) {
            if (!$child->evaluate($attrs)) {
                return false;
            }
        }

        return true;
    }
}

final class FilterOr extends FilterExpr
{
    /**
     * @param FilterExpr[] $children
     */
    public function __construct(private array $children)
    {
    }

    public function evaluate(OrderedAttrs $attrs): bool
    {
        foreach ($this->children as $child) {
            if ($child->evaluate($attrs)) {
                return true;
            }
        }

        return false;
    }
}

final class FilterNot extends FilterExpr
{
    public function __construct(private FilterExpr $child)
    {
    }

    public function evaluate(OrderedAttrs $attrs): bool
    {
        return !$this->child->evaluate($attrs);
    }
}

final class ItemMatcher
{
    /**
     * @param string[] $parts
     */
    public function __construct(
        public string $kind,
        public array $parts = [],
        public bool $leadingStar = false,
        public bool $trailingStar = false,
    ) {
    }
}

final class FilterItem extends FilterExpr
{
    public function __construct(
        private string $attr,
        private string $op,
        private string $value,
        private ItemMatcher $matcher,
    ) {
    }

    public function evaluate(OrderedAttrs $attrs): bool
    {
        $actual = $attrs->get($this->attr);

        if ($this->op === '=') {
            if ($this->matcher->kind === 'presence') {
                return $attrs->contains($this->attr);
            }

            if ($actual === null) {
                return false;
            }

            if ($this->matcher->kind === 'wildcard') {
                return wildcardMatches(
                    $this->matcher->parts,
                    $this->matcher->leadingStar,
                    $this->matcher->trailingStar,
                    $actual,
                );
            }

            return $actual === $this->value;
        }

        if ($actual === null) {
            return false;
        }

        if ($this->op === '~=') {
            return levenshtein($actual, $this->value) <= 2;
        }

        if ($this->op === '>=') {
            return $actual >= $this->value;
        }

        if ($this->op === '<=') {
            return $actual <= $this->value;
        }

        throw new LdapFilterError("unsupported operator: {$this->op}");
    }
}

final class OrderedAttrs
{
    /**
     * @var array<int, array{0: string, 1: ?string}>
     */
    private array $items = [];

    public function add(string $key, ?string $value): void
    {
        $this->items[] = [$key, $value];
    }

    public function contains(string $key): bool
    {
        foreach ($this->items as [$candidate, $_]) {
            if ($candidate === $key) {
                return true;
            }
        }

        return false;
    }

    public function get(string $key): ?string
    {
        foreach ($this->items as [$candidate, $value]) {
            if ($candidate === $key) {
                return $value;
            }
        }

        return null;
    }

    /**
     * @return array<int, array{0: string, 1: ?string}>
     */
    public function items(): array
    {
        return $this->items;
    }
}

final class FilterParser
{
    public function __construct(
        private string $text,
    ) {
    }

    public function parse(): FilterExpr
    {
        $expr = $this->parseFilter();
        if ($this->pos !== strlen($this->text)) {
            throw $this->error('unexpected trailing characters');
        }

        return $expr;
    }

    private int $pos = 0;

    private function parseFilter(): FilterExpr
    {
        $this->expect('(');
        $current = $this->peek();
        if ($current === null) {
            throw $this->error('parenthesis mismatch');
        }

        if ($current === '&') {
            $this->pos++;
            $children = $this->parseSubfilters();
            if ($children === []) {
                throw $this->error('and operator requires filters');
            }
            $this->expect(')');
            return new FilterAnd($children);
        }

        if ($current === '|') {
            $this->pos++;
            $children = $this->parseSubfilters();
            if ($children === []) {
                throw $this->error('or operator requires filters');
            }
            $this->expect(')');
            return new FilterOr($children);
        }

        if ($current === '!') {
            $this->pos++;
            $child = $this->parseFilter();
            if ($this->peek() === '(') {
                throw $this->error('not operator has more than one filter');
            }
            $this->expect(')');
            return new FilterNot($child);
        }

        $item = $this->parseItem();
        $this->expect(')');
        return $item;
    }

    /**
     * @return FilterExpr[]
     */
    private function parseSubfilters(): array
    {
        $children = [];
        while ($this->peek() === '(') {
            $children[] = $this->parseFilter();
        }
        return $children;
    }

    private function parseItem(): FilterItem
    {
        $start = $this->pos;
        while (true) {
            $current = $this->peek();
            if ($current === null) {
                throw $this->error('parenthesis mismatch');
            }
            if ($current === ')') {
                break;
            }
            $this->pos++;
        }

        $item = substr($this->text, $start, $this->pos - $start);
        if (!preg_match('/\A([^~=><]+)(~=|>=|<=|=)(.+)\z/s', $item, $matches)) {
            throw $this->error('error in item syntax', $start);
        }

        $attr = $matches[1];
        $op = $matches[2];
        $rawValue = $matches[3];
        [$value, $matcher] = parseItemValue($rawValue, $start);
        return new FilterItem($attr, $op, $value, $matcher);
    }

    private function peek(): ?string
    {
        if ($this->pos >= strlen($this->text)) {
            return null;
        }

        return $this->text[$this->pos];
    }

    private function expect(string $expected): void
    {
        $actual = $this->peek();
        if ($actual !== $expected) {
            if ($actual === null) {
                throw $this->error("expected " . var_export($expected, true) . ', found EOF');
            }
            throw $this->error("expected " . var_export($expected, true) . ', found ' . var_export($actual, true));
        }
        $this->pos++;
    }

    private function error(string $message, ?int $position = null): LdapFilterError
    {
        if ($position === null) {
            $position = $this->pos;
        }
        return new LdapFilterError(sprintf('%s at position %d', $message, $position));
    }
}

function parseFilter(string $text): FilterExpr
{
    return (new FilterParser($text))->parse();
}

/**
 * @return array{0: string, 1: ItemMatcher}
 */
function parseItemValue(string $raw, int $position): array
{
    if ($raw === '*') {
        return ['*', new ItemMatcher('presence')];
    }

    $decoded = '';
    $segments = [];
    $current = '';
    $sawWildcard = false;
    $len = strlen($raw);

    for ($i = 0; $i < $len;) {
        $ch = $raw[$i];
        if ($ch === '\\') {
            if ($i + 2 >= $len) {
                throw new LdapFilterError(sprintf('incomplete escape sequence at position %d', $position));
            }
            $decodedChar = decodeHexEscape($raw[$i + 1], $raw[$i + 2], $position);
            $decoded .= $decodedChar;
            $current .= $decodedChar;
            $i += 3;
            continue;
        }

        if ($ch === '*') {
            $sawWildcard = true;
            $decoded .= '*';
            $segments[] = $current;
            $current = '';
            $i++;
            continue;
        }

        $decoded .= $ch;
        $current .= $ch;
        $i++;
    }

    if (!$sawWildcard) {
        return [$decoded, new ItemMatcher('exact')];
    }

    $segments[] = $current;
    return [
        $decoded,
        new ItemMatcher(
            'wildcard',
            $segments,
            str_starts_with($raw, '*'),
            str_ends_with($raw, '*'),
        ),
    ];
}

function decodeHexEscape(string $high, string $low, int $position): string
{
    if (!ctype_xdigit($high) || !ctype_xdigit($low)) {
        throw new LdapFilterError(sprintf('invalid escape sequence: \\%s%s at position %d', $high, $low, $position));
    }

    return chr(hexdec($high . $low));
}

/**
 * @param string[] $parts
 */
function wildcardMatches(array $parts, bool $leadingStar, bool $trailingStar, string $actual): bool
{
    $partCount = count($parts);
    if ($partCount === 0) {
        return $actual === '';
    }

    $start = 0;
    $end = $partCount - 1;
    $offset = 0;

    if (!$leadingStar) {
        $first = $parts[0];
        if (!str_starts_with($actual, $first)) {
            return false;
        }
        $offset = strlen($first);
        $start = 1;
    }

    if (!$trailingStar) {
        $last = $parts[$end];
        if ($last !== '') {
            $lastPos = strrpos($actual, $last);
            if ($lastPos === false || $lastPos + strlen($last) !== strlen($actual)) {
                return false;
            }
        }
        $end--;
    }

    for ($i = $start; $i <= $end; $i++) {
        $part = $parts[$i];
        if ($part === '') {
            continue;
        }

        $found = strpos($actual, $part, $offset);
        if ($found === false) {
            return false;
        }

        $offset = $found + strlen($part);
    }

    return true;
}

function inspectAttrs(OrderedAttrs $attrs): string
{
    $parts = [];
    foreach ($attrs->items() as [$key, $value]) {
        $parts[] = formatKey($key) . formatValue($value);
    }
    return '{' . implode(', ', $parts) . '}';
}

function formatKey(string $key): string
{
    if (isRubySymbolName($key)) {
        return $key . ': ';
    }
    return json_encode($key, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES) . ' => ';
}

function formatValue(?string $value): string
{
    if ($value === null) {
        return 'nil';
    }
    return json_encode($value, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
}

function isRubySymbolName(string $key): bool
{
    return (bool) preg_match('/\A[A-Za-z_][A-Za-z0-9_]*\z/', $key);
}

/**
 * @return array{0: string|null, 1: string|null}
 */
function parseArgs(array $argv): array
{
    $options = [
        'filter' => null,
        'input' => null,
        'format' => 'auto',
    ];
    $positionals = [];

    $count = count($argv);
    for ($i = 1; $i < $count; $i++) {
        $arg = $argv[$i];

        if ($arg === '--help' || $arg === '-h') {
            printUsage();
            exit(0);
        }

        if ($arg === '--jit' || $arg === '--no-jit') {
            continue;
        }

        if ($arg === '--filter' || $arg === '--input' || $arg === '--format') {
            if ($i + 1 >= $count) {
                throw new InvalidArgumentException("missing value for {$arg}");
            }
            $value = $argv[++$i];
            if ($arg === '--filter') {
                $options['filter'] = $value;
            } elseif ($arg === '--input') {
                $options['input'] = $value;
            } else {
                $options['format'] = $value;
            }
            continue;
        }

        if (str_starts_with($arg, '--filter=')) {
            $options['filter'] = substr($arg, 9);
            continue;
        }
        if (str_starts_with($arg, '--input=')) {
            $options['input'] = substr($arg, 8);
            continue;
        }
        if (str_starts_with($arg, '--format=')) {
            $options['format'] = substr($arg, 9);
            continue;
        }

        if (str_starts_with($arg, '--')) {
            throw new InvalidArgumentException("unsupported option: {$arg}");
        }

        $positionals[] = $arg;
    }

    if ($options['filter'] === null && isset($positionals[0])) {
        $options['filter'] = $positionals[0];
    }
    if ($options['input'] === null && isset($positionals[1])) {
        $options['input'] = $positionals[1];
    }

    return [$options['filter'], $options['input'], $options['format']];
}

function printUsage(): void
{
    fwrite(STDOUT, "Usage: php ldap_filter.php [options] FILTER INPUT\n");
}

function runWithIo(array $argv): int
{
    [$filter, $inputPath, $format] = parseArgs($argv);
    if ($filter === null || $inputPath === null) {
        throw new InvalidArgumentException('filter and input path are required');
    }

    $startNs = monotonicNs();
    fwrite(STDERR, phaseLine('boot', $startNs, monotonicNs()) . "\n");

    $format = selectFormat($format, $inputPath);
    $expr = parseFilter($filter);

    fwrite(STDERR, phaseLine('ready', $startNs, monotonicNs()) . "\n");

    processInput($inputPath, $format, $expr);

    fwrite(STDERR, phaseLine('done', $startNs, monotonicNs()) . "\n");
    return 0;
}

function selectFormat(string $format, string $inputPath): string
{
    if ($format !== 'auto') {
        return $format;
    }

    return detectFormat($inputPath);
}

function detectFormat(string $inputPath): string
{
    $base = basename($inputPath);
    if (str_ends_with($base, '.csv') || str_ends_with($base, '.csv.xz')) {
        return 'csv';
    }
    if (str_ends_with($base, '.ltsv') || str_ends_with($base, '.ltsv.xz')) {
        return 'ltsv';
    }

    $firstLine = withInputHandle($inputPath, static function ($handle): ?string {
        while (($line = fgets($handle)) !== false) {
            if (trim($line) !== '') {
                return $line;
            }
        }
        return null;
    });

    if ($firstLine !== null && str_contains($firstLine, "\t")) {
        return 'ltsv';
    }

    return 'csv';
}

function processInput(string $inputPath, string $format, FilterExpr $expr): void
{
    if ($format === 'csv') {
        eachCsvAttrs($inputPath, static function (OrderedAttrs $attrs) use ($expr): void {
            if ($expr->evaluate($attrs)) {
                fwrite(STDOUT, inspectAttrs($attrs) . "\n");
            }
        });
        return;
    }

    if ($format === 'ltsv') {
        eachLtsvAttrs($inputPath, static function (OrderedAttrs $attrs) use ($expr): void {
            if ($expr->evaluate($attrs)) {
                fwrite(STDOUT, inspectAttrs($attrs) . "\n");
            }
        });
        return;
    }

    throw new InvalidArgumentException("unsupported format: {$format}");
}

/**
 * @param callable(OrderedAttrs): void $callback
 */
function eachCsvAttrs(string $inputPath, callable $callback): void
{
    withInputHandle($inputPath, static function ($handle) use ($callback): void {
        $headers = fgetcsv($handle);
        if ($headers === false) {
            return;
        }

        if ($headers !== [] && isset($headers[0])) {
            $headers[0] = preg_replace('/^\xEF\xBB\xBF/', '', (string) $headers[0]);
        }

        while (($row = fgetcsv($handle)) !== false) {
            if ($row === [null]) {
                continue;
            }

            $attrs = new OrderedAttrs();
            $headerCount = count($headers);
            for ($i = 0; $i < $headerCount; $i++) {
                $value = $row[$i] ?? '';
                $attrs->add((string) $headers[$i], $value === null ? '' : (string) $value);
            }

            $callback($attrs);
        }
    });
}

/**
 * @param callable(OrderedAttrs): void $callback
 */
function eachLtsvAttrs(string $inputPath, callable $callback): void
{
    withInputHandle($inputPath, static function ($handle) use ($callback): void {
        while (($line = fgets($handle)) !== false) {
            $line = rtrim($line, "\r\n");
            if ($line === '') {
                continue;
            }

            $attrs = parseLtsvLine($line);
            $callback($attrs);
        }
    });
}

function parseLtsvLine(string $line): OrderedAttrs
{
    $attrs = new OrderedAttrs();
    foreach (explode("\t", $line) as $field) {
        if ($field === '') {
            continue;
        }

        $separator = strpos($field, ':');
        if ($separator === false) {
            continue;
        }

        $key = substr($field, 0, $separator);
        $value = substr($field, $separator + 1);
        $attrs->add($key, unescapeLtsvValue($value));
    }

    return $attrs;
}

function unescapeLtsvValue(string $text): string
{
    return str_replace(['\\\\', '\t', '\n', '\r'], ['\\', "\t", "\n", "\r"], $text);
}

/**
 * @template T
 * @param callable(string): T $callback
 * @return T
 */
function withInputHandle(string $inputPath, callable $callback)
{
    if (str_ends_with($inputPath, '.xz')) {
        $cmd = 'xz -dc ' . escapeshellarg($inputPath);
        $descriptors = [
            0 => ['pipe', 'r'],
            1 => ['pipe', 'w'],
            2 => ['pipe', 'w'],
        ];
        $pipes = [];
        $process = proc_open($cmd, $descriptors, $pipes);
        if (!is_resource($process)) {
            throw new RuntimeException("failed to open xz process for {$inputPath}");
        }

        fclose($pipes[0]);
        try {
            return $callback($pipes[1]);
        } finally {
            if (isset($pipes[1]) && is_resource($pipes[1])) {
                fclose($pipes[1]);
            }
            $stderr = '';
            if (isset($pipes[2]) && is_resource($pipes[2])) {
                $stderr = stream_get_contents($pipes[2]) ?: '';
                fclose($pipes[2]);
            }
            $exitCode = proc_close($process);
            if ($exitCode !== 0) {
                throw new RuntimeException(sprintf('xz failed for %s: %s', $inputPath, trim($stderr)));
            }
        }
    }

    $handle = fopen($inputPath, 'rb');
    if ($handle === false) {
        throw new RuntimeException("failed to open {$inputPath}");
    }

    try {
        return $callback($handle);
    } finally {
        fclose($handle);
    }
}

function phaseLine(string $phase, int $startNs, int $nowNs): string
{
    $elapsedNs = $nowNs - $startNs;
    return sprintf('phase=%s t=%d elapsed_ns=%d', $phase, $elapsedNs, $elapsedNs);
}

function monotonicNs(): int
{
    return hrtime(true);
}

function main(array $argv): int
{
    try {
        return runWithIo($argv);
    } catch (Throwable $e) {
        fwrite(STDERR, $e->getMessage() . "\n");
        return 1;
    }
}

if (PHP_SAPI === 'cli' && realpath($argv[0] ?? '') === __FILE__) {
    exit(main($argv));
}
